src/main/flight/alt_hold.c

```diff
@@ -0,0 +1,241 @@
+
+    // Some people may not like throttle being able to change the target altitude, because:
+    // - with throttle adjustment, hitting the switch won't always hold altitude if throttle is bumped
+    // - eg if the throttle is bumped low accidentally, quad will start descending.
+    // On the plus side:
+    // - the pilot has control nice control over altitude, and the deadband is wide
+    // - Slow controlled descents are possible, eg for landing
+    // - fine-tuning height is possible, eg if there is slow sensor drift
+    // - to keep the craft stable, throttle must be in the deadband, making exits smoother
+
+    const float rcThrottle = rcCommand[THROTTLE];
+
+    const float lowThreshold = 0.5f * (positionConfig()->hover_throttle + PWM_RANGE_MIN); // around 1150
+    const float highThreshold = 0.5f * (positionConfig()->hover_throttle + PWM_RANGE_MAX); // around 1500
+    
+    float throttleAdjustmentFactor = 0.0f;
+    if (rcThrottle < lowThreshold) {
+        throttleAdjustmentFactor = scaleRangef(rcThrottle, PWM_RANGE_MIN, lowThreshold, -1.0f, 0.0f);
+    } else if (rcThrottle > highThreshold) {
+        throttleAdjustmentFactor = scaleRangef(rcThrottle, highThreshold, PWM_RANGE_MAX, 0.0f, 1.0f);
+    }
+
+    // if failsafe is active, and we get here, we are in landing mode, so descend at configured max rate
+    if (failsafeIsActive()) {
+        throttleAdjustmentFactor = -1.0f;
+    }
+
+    altHoldState.targetAltitudeDelta = throttleAdjustmentFactor * altholdConfig()->altHoldTargetAdjustRate * taskIntervalSeconds;
+    // if taskRate is 100Hz, default adjustRate of 100 adds/subtracts 1m every second, or 1cm per task run, at full stick position
+    altHoldState.targetAltitudeCm += altHoldState.targetAltitudeDelta;
+}
+
+void altHoldUpdate(void)
+{
+    // things that always need to happen in the background should go here
+    altHoldState.measuredAltitudeCm = getAltitude();
+    DEBUG_SET(DEBUG_ALTHOLD, 1, lrintf(altHoldState.measuredAltitudeCm));
+
+    altHoldProcessTransitions(); // initialise values, don't start without altitude data
+
+    // exit unless altHold is active
+    if (!altHoldState.isAltHoldActive) {
+        const uint8_t altitudeDataAvailable = isAltitudeAvailable();
+        DEBUG_SET(DEBUG_ALTHOLD, 0, altitudeDataAvailable);
+        return;
+    }
+
+    // if active, and we lose altitude readings (lose GPS 3D fix for example)
+    // if the last measured altitude was below target, we will climb indefinitelym, and vice versa
+    // we can set current altitude to target, making PIDs zero, and must force zero iTerm
+    // quad then gets the un-modified hover value and may climb or descend, but this should be slow
+    // user will havbe NO control user needs a warning (how? esp LOS) and should exit
+    if (!isAltitudeAvailable()) {
+        altHoldState.measuredAltitudeCm = altHoldState.targetAltitudeCm;
+        // force integral to zero or will accumulate
+        simplePid.integral = 0.0f;
```

Comment:
This will cause jump in throttle (I-term is roughly estimating necessary hover throttle). It may be better to freeze i-term (and zero error should freeze it)

Reply:
Hmm.
It depends I guess on whether we anticipate a hardware fault of this kind to be temporary or permanent.  Most likely it is permanent.
While iTerm does, in a stable hover, correct for battery sag etc, it can be transiently high if there was a transient reason for it to be high.  So then the quad could climb if we hold the last value.
Now I guess that on a fresh battery it could climb also with iTerm forced to zero, but we can't fix that.
In this case a small throttle jump when returning iTerm to zero is unlikely to matter much.
The real question is whether we want to ensure that the quad descends, rather than climbs.  We could perhaps force iTerm to a small negative value eg -20 which would tend to favour descent?

Reply:
On reflection I simplified the code so that we just exit the hold if we lose the altitude data after initiation of the hold.
It is not likely to happen, and the pilot should detect the failure visually.
If it happens in failsafe landing mode, the previous behaviour applies of constant throttle (default is 0, currently)